# HBV MUTATIONAL ANALYSIS

## BACKGROUND

This method has been developed to carry out the screening of mutations on FASTA files obtained from the sequencing of hepatitis B virus (HBV) samples received by the Blood Borne Virus unit (BBVU). 

Consensus nucleotide sequences generated by PCR amplification of two different regions and Sanger sequencing are used for the analysis of:

- **SP amplicon**, covering a region between nucleotides 892 and 1906 (*numbering AY128092.1 Polymerase*), used to analyse the substitutions in the protein Surface and Polymerase (RT domain). 
- **XC amplicon**, covering a region between nucleotides 1656 and 2402 (*numbering AY128092.1 complete genome*), used to analyse the substitutions in the protein X and Precore. 
   
The script takes FASTA files containing the consensus nucleotides sequences and a CSV file consisting of a sample list with the genotypes for the samples (if available) and performs **pair alignments** between each sequence and a specific reference sequence provided. The sequences from XC amplicons are aligned to the reference sequence [LC360507.1_WG_xprecore.fas](refseqs/LC360507.1_WG_xprecore.fas) whereas the ones generated from SP amplicons are aligned to the corresponding reference sequence ([*"_polymerase_surface"* found in directory refseqs](refseqs)), depending of the genotype indicated for that sequence in the CSV file. 

An initial nucleotide alignment is used to **trim the sequences at the lenght of the region of interests** (X region, Precore, Polymerase-RT or Surface). Then, the sequences are analysed for **potential frame-shifts**. As the initial alignment is trimmed using the coordinates for the reference sequence, this one is in frame if any gap is removed from the nucleotide sequence. Three different frames are evaluated for the consensus sequence after removing gaps from the nucleotide sequences, then aligned to the original trimmed ungapped reference sequence and the resulting alignment is scored for similarities. If the first frame (original one) gets the higher score, the alignment goes through the analysis of mutations. Otherwise the sequence is reported as frame-shifted and no further analysis is performed.  

A **codon alignment** is carried out in a next step. Gaps in the nucleotide alignment are removed and the resulting alignment is translated to amino acids and aligned. The protein alignment is then used to *translate the nucleotide nucleotides back*. Each *amino acid* from the protein alignment corresponds to a *codon* from the ungapped nucleotide alignment, therefore every position in the amino acid alignment matches three positions in a nucleotide sequence. In addition, *gaps (-)* in the protein alignment should be equivalent to a *triplet of hyphens (---)*. These principles are used to re-organise the nucleotide sequences into a new nucleotide alignment.

The codons from the new nucleotide alignment are **translated to amino acids** and they are stored in a list, keeping the exact order they have in the alignment. A second list is used to store the amino acids of the reference and the consensus sequences. The two lists are used to explore potential **insertions** in the *consensus* sequence, which are detected by the presence of the notation **"---"** at any codon of the *reference* sequence. The amino acid involved in the insertion and the position are annotated and them removed from the list as well as the codon contributing to the insertions. In this way, the numbering for the codons and amino acids in the lists are the same as the reference sequence used for the alignment. 

In a final step, the corrected list of codons and amino acids are used to **scan for variations** between the *reference* sequence and the *consensus* sequence. Any change that differs from the reference sequence or from [substitutions considered as common on HBV strains](Amino_Acid_Templates_BBVU) are annotated. 

All the changes found are then recorded in a excel file that the user can explore. 

## REQUIREMENT
The main packages and Python libraries used for this method are:
- Python 3.6.1
- Biopython 1.78
- Pandas 1.1.5
- Openpyxl 3.0.9
- Clustalw 2.1

The easiest way of installing these software packages and other additional dependencies is by means of using [Anaconda distribution](https://www.anaconda.com/products/individual). Visit https://docs.anaconda.com/anaconda/install/ for installation details. 

With [git](https://git-scm.com/) and Anaconda already installed in the computer, open a terminal and navigate to the **main directory** where all the Sequencing runs are stored and open the terminal. 

Then enter the follwoing command in your terminal:

`git clone https://github.com/juanledesma78/hbv_mutational_analysis.git`

A new subdirectory called **hbv_mutational_analysis**, containing all the files from the repository, is created in the main directory. A file called **environment.yml**, which contains the list of all the software and dependencies needed, should be found in the recently created subdirectory. This file can be used to create a new environment called **hbv_mutations** in Anaconda. Before creating the environment, replace the line */home/phe.gov.uk/juan.ledesma/miniconda3/* in the *prefix* of the yml file with the path to where conda is installed on your computer. A simple way to get that path is typing in a terminal:

`conda env list`

which will return the path to base or where conda is located 
```
# conda environments:
#
base                  *  /home/phe.gov.uk/another.user/miniconda3

```
Once the prefix has been modified (i.e.*prefix: /home/phe.gov.uk/another.user/miniconda3/envs/hbv_mutations*), use the following command to create the new environment with all the software listed in the yml file

```
conda env create -f environment.yml
```

## USAGE

While the whole repository includes several folders and files, the ones **required** to make the method work are:
- **refseqs**, which contains 10 FASTA files with nucleotide sequences for the analysis of polymerase and surface for each genotype (A-J) as well as a single FASTA file for the analysis of Precore and X region. These files were used to align the consensus sequences provided from the runs.   

  * AY128092.1_Genotype_A_polymerase_surface.fas
  * AY167089.1_Genotype_B_polymerase_surface.fas
  * KM999990.1_Genotype_C_polymerase_surface.fas
  * X65257.1_Genotype_D_polymerase_surface.fas
  * X75657.1_Genotype_E_polymerase_surface.fas
  * KX264497.1_Genotype_F_polymerase_surface.fas
  * AF160501.1_Genotype_G_polymerase_surface.fas
  * AY090457.1_Genotype_H_polymerase_surface.fas
  * AF241411.1_Genotype_I_polymerase_surface.fas
  * AB486012.1_Genotype_J_polymerase_surface.fas
  * LC360507.1_WG_xprecore.fas

- **Sample_list_template.csv**, an empty CSV file containing the headers *Seq ID*	and *Genotype*,

- **hbv_mutational_analysis.py**, the script to run the analysis,

- **mutation_utilities.py**, a module with functions used in the analysis,

They should be transferred to the **main directory** where all the Sequencing runs are stored, so the script **hbv_mutational_analysis.py** has direct access to the run folders. 

The ***INPUTS*** the method uses to perform the mutational analysis are: 

- **FASTA files** stored in a subdirectory called **Consensus** within the run directory (i.e E1647/Consensus). These FASTA files contain nucleotide sequences generated by the sequencing of the PCR **SP** and **XC** amplicons , described above. The identifiers or headers of the FASTA files must be named following the following pattern: ***tubeNo_molisID*** followed by either ***sp*** or ***xc*** depending of the PCR amplicon used to generate the sequence. Note that tube number and molis ID are separated by underscore and no other separator should be used. If the user wants to add descriptive information to the FASTA header, underscore (_) should be avoided using instead other characters for this purpose (i.e 26_H214561122-EXTRACTED-USING-MAGNAPURExc). 

- A **Sample_list.csv**. The file **Sample_list_template.csv** provided with the repository can be used to insert the sample IDs to be analysed to the column *Seq ID*. Note that the sample id **MUST** match the FASTA header after removing ***sp*** or ***xc***. The same sample list used for the extraction could be used for the column *Seq ID* as long as the IDs follow the same naming rule described above. If information for the genotype of the samples is available for the analysis of polymearse/surface, it should be added to the column *Genotype*. The first character in the row, written in capital letter, **MUST** correspond to any of the genotypes describe for HBV (A-J) but other comments are allowed after this first one (i.e. F1/F1 or D1 by core). N/A for sequences where the genotype is unknown is permited as well. Once the CSV file is filled with the information required, it **MUST** be saved in the same subdirectory where the FASTA files are (i.e E1647/Consensus) under the name **Sample_list.csv**.

Before running the script, the conda environment **MUST** be activated. From the main directory containing the Sequencing RUNS, type in a terminal

`conda activate hbv_mutations`


Without changing directory the analysis can be started now. The script **hbv_mutational_analysis.py** takes one argument (**RUNID**), which should be added after -dir. In case of doubt, the user can consult some help to run the script by typing:

`python3 hbv_mutational_analysis.py --help`

To start the analysis of a run, type

`python3 hbv_mutational_analysis.py -dir <run id, i.e. E1647>`

A message will be displayed with the links to where the results are availabe If the process has successfully been complete.

The ***OUTPUTS*** of the method are:

- a **txt file** (i.e. *E1647_hbvma_v1.0_log.txt*), which corresponds to a **log** with a summary of the *inputs*, *outputs* and potential *warnings* of the analysis. It is **strongly recommended** for the user to have a look of it before checking the results, in case some samples have not been analysed due to omission of their Ids in the CVS file or the absence of fasta files for the analysis.  

- a **report in XLSX format** (i.e. *Results_E1647_hbvma_v1.0.xlsx*) that contains a first tab called *All* which includes a list with all the sample IDs and genotypes provided in *Sample_list.csv* alongside the results after analysing the 4 genenomic regions. NA is used when a FASTA file is not found in the directory RUNID/Consensus (either becuase that amplicon has not been sequenced or the FASTA file has not been not provided).  Ten additional tabs (A-J) are used to record the results according to the genotype of the samples and a final tab called *CORE*  is used to record the results available for genes X and Precore. Two final tabs, *Numbering* and *Extended_IUPAC_aa_code*, are used to display the accession numbers for the numbering of the substitution in the results and the symbols to refer the 20 standard amino acids as well as other potential combinations of ambiguous amino acids, respectively.  

- a directory **alignmnets** that contains the alignments between each consensus sequence and the reference sequence used for the analysis of *polymearseRT*, *surface*, *xregion* and *precore*, information provided in the excel file. The numbering of the substutions recorded in the report are based on the reference sequence included in each alignmnet. An additional subdirectory called **raw_trimmed** is used to store the initial nucelotide alignmets after trimming for the length of the four genomic regions. These alignments are used for the script to check potential frame-shifts in the sequences. Samples with the label *"Sequence NOT in frame"* in any of the genomic regions will find their corresponding alignmnet **ONLY in this subdirectory**, in case the user wants to check the positions involved in the frame-shift.   

## TESTING

Three clinical runs ([E1647](testing/E1647), [E1651](testing/E1651) and [E1665](testing/E1665)) has been used so far to test the performance of the currenrt method. The results of the method hbv_mutational_analysis can be checked in [Results_E1647_hbvma_v1.0.xlsx](testing/E1647/Results_E1647_hbvma_v1.0.xlsx), [Results_E1651_hbvma_v1.0.xlsx](testing/E1647/Results_E1651_hbvma_v1.0.xlsx) and [Results_E1665_hbvma_v1.0.xlsx](testing/E1647/Results_E1665_hbvma_v1.0.xlsx) and compared to the result of the manual method Results_E1647.xlsx](testing/E1647/Results_E1647.xlsx), [Results_E1651.xlsx](testing/E1647/Results_E1651.xlsx) and [Results_E1665.xlsx](testing/E1647/Results_E1665.xlsx). 

The current method displays all combined substitution generated by ambiguous nucleotides (i.e S207N/S in Surface, run [E1647](testing/E1647/Results_E1647_hbvma_v1.0.xlsx)). Ambiguous amino acids such as **B** (N/D, Asparagine or Aspartic acid), **Z** (E/Q,	Glutamic acid or Glutamine) or **J** (L/I, Leucine or Isoleucine) may be recorded in the report, so the user should be aware of that and check the tab *Extended_IUPAC_aa_code* in the EXCEL report for additional information.

The analysis of the Polymerase (RT domain) in the new method covers from the first amino acid to amino acid 269 (position 73 to 250 in the manual method) so any change differing from the reference sequence is recorded for this region. Althought the analysis of the X region covers the same partial fragment in both methods, the new method does not display the first amino acid of the amplicon. This information is used in the script to perform some steps and it may be included in future versions. The analysis for precore and surface covers the same regions in both methods. 

The new method includes **genotypes A-J** so any sample belonging to any of these genotypes can be analysed by this method. 

[Another run based on modified sequences](testing/final) was used to test the performance of the new method for: 
- the **analysis of mutations in genotypes G, J and I**
- detection of **frameshifts**
- detection of the **severals indels** in the sequence 05_H214460979sp
    * deletion of amino acid R at position 18
    * insertion of amino acid R at position 60
    * deletion of amino acids L and S at position 108 and 109 respectively
    * insertion of amino acid P at position 140
- generation of the report 

The results are available in [Results_final_hbvma_v1.0.xlsx](testing/final/Results_final_hbvma_v1.0.xlsx). 

## LIMITATIONS

Clustal does **NOT** deal with **ambiguous amino acids** *(BZJUO)* or *stop codons* (*) when doing **protein alignments** (step needed for **codon alignment**). In the script these characters are replaced with X to allow the alignment of amino acids. Otherwise clustal considers them as gaps and the amino acid sequences could be misaligned. 
Other alignmers such as [MAFFT](https://mafft.cbrc.jp/alignment/software/) or [MUSCLE](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-5-113) may be considered for future versions. 

The method has been developed in a linux machine using Ubuntu 18.04. For Windows user, this version has not been fully tested yet.  

## CONTACT DETAILS
Juan Ledesma

VRD Bioinformatics 

UK Health Security Agency

juan.ledesma@phe.gov.uk 

February 2022

